.devcontainer/devcontainer.json

```diff
@@ -1,36 +1,31 @@
-// For format details, see https://aka.ms/vscode-remote/devcontainer.json or this file's README at:
-// https://github.com/microsoft/vscode-dev-containers/tree/v0.195.0/containers/cpp
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/cpp
 {
 	"name": "C++",
 	"build": {
-		"dockerfile": "Dockerfile",
-		// Update 'VARIANT' to pick an Debian / Ubuntu OS version: debian-11, debian-10, debian-9, ubuntu-21.04, ubuntu-20.04, ubuntu-18.04
-		// Use Debian 11, Debian 9, Ubuntu 18.04 or Ubuntu 21.04 on local arm64/Apple Silicon
-		"args": { "VARIANT": "debian-11" }
+		"dockerfile": "Dockerfile"
 	},
-	"runArgs": [ "--cap-add=SYS_PTRACE", "--security-opt", "seccomp=unconfined"],
-		
+
+	// Features to add to the dev container. More info: https://containers.dev/features.
+	// "features": {},
+
 	// Configure tool-specific properties.
 	"customizations": {
 		// Configure properties specific to VS Code.
 		"vscode": {
-			// Set *default* container specific settings.json values on container create.
 			"settings": {},
-			
-			// Add the IDs of extensions you want installed when the container is created.
 			"extensions": [
-				"ms-vscode.cpptools"
```

Comment:
This breaks the OOBE for this example.

Recommend `ms-vscode.cpptools-extension-pack` instead, otherwise the user will get a toast notification for "C++ tools" that wants to install the extension pack.

This should require a correlated change in the `README.md` for this bullet point

Reply:
It actually doesn't if you test out the example. The image now includes this extension as explained in the README updates.  This part of the new image metadata capability for Dev Containers. 

We could, however, flip to the extension pack.  Right now its CMake and C++: https://github.com/devcontainers/images/blob/main/src/cpp/.devcontainer/devcontainer.json#L14-L15

Reply:
So, if I understand this correctly then, the container image in the `FROM` line of the `Dockerfile` is constructed from a dev container spec at that link.

Do you have a link to the documentation for the new image metadata capability for Dev Containers? I'm not familiar with it.

When I tested the example, I got a VSCode notification asking to install C/C++ extensions. After I manually added back `ms-vscode.cpptools` I expected that to go away, but it didn't, which is when I realized that that notification goes away with the extension pack instead. That might be worth improving on for the C++ extension (@AlexandraKemperMS).

I let an extra variable sneak into my test. :)

FWIW, I'm not crazy about hiding that implementation detail. If we expect people to learn something other than "I guess it does work." from these repositories, they may not immediately follow the leap towards the extensions being installed in the base image via a different devcontainer.json file. For me, who has some experience with using them, it really did feel like something critically important was just missing.

Is there some advantage to putting those in the base image that I'm not understanding?

Reply:
Make sure you are on the latest version of the extension. This is a new capability. I can double check. 

The link to the docs is in the README. https://devcontainers.github.io/implementors/spec/ has the details on processing and https://devcontainers.github.io/implementors/json_reference/ annotates which properties are supported. 

The goal is to prevent config drift by keeping the settings needed for the image with the image. It also allows mass config updates across many repos at once which has been a common ask. Finally it simplifies reuse since you can just ref an image directly if you prefer.